### PR TITLE
lerp LinearGradient

### PR DESCRIPTION
**Pull request by xster**
To use for #9321

### CODE REVIEW

File: packages/flutter/lib/src/painting/box_painter.dart [hunk 1]
```diff
@@ -679,6 +682,46 @@ class LinearGradient extends Gradient {
...
+    if (a == null && b == null)
+      return null;
+    if (a == null)
+      return b.scale(t);
+    if (b == null)
+      return a.scale(1.0 - t);
+    // Interpolation is only possible when the lengths of colors and stops are
+    // the same or stops is null for one.
+    assert(a.colors.length == b.colors.length);
+    assert(a.stops == null || b.stops == null || a.stops.length == b.stops.length);
```
Comment (abarth) [inline packages/flutter/lib/src/painting/box_painter.dart]: We could probably figure out a lerp for these cases, but maybe this is fine for now.

**Comment by abarth** [comment]
LGTM

File: packages/flutter/lib/src/painting/box_painter.dart (hunk 1 above)
Comment (xster) [inline packages/flutter/lib/src/painting/box_painter.dart]: ya, we can't really mix them. We'd probably have to refactor BoxDecoration to take a list of gradients and scale them separately.

**Comment by Hixie** [comment]
I added some comments on the other PR that apply here.

File: packages/flutter/lib/src/painting/box_painter.dart [hunk 2]
```diff
@@ -707,18 +707,25 @@ class LinearGradient extends Gradient {
...
-      interpolatedColors.add(Color.lerp(a.colors.elementAt(i), b.colors.elementAt(i), t));
-    }
+    for (int i = 0; i < a.colors.length; i++) 
+      interpolatedColors.add(Color.lerp(a.colors[i], b.colors[i], t));
+    List<double> interpolatedStops;
+    if (a.stops != null && b.stops != null) {
+      for (int i = 0; i < a.stops.length; i++)
+        interpolatedStops.add(ui.lerpDouble(a.stops[i], b.stops[i], t));
+    } else 
+      interpolatedStops = a.stops ?? b.stops;
```
Comment (Hixie) [inline packages/flutter/lib/src/painting/box_painter.dart]: nit: yeah, sorry, the style guide is vague (or even technically wrong) about this. the prevailing style is to either have braces on both sides of the else or have them or neither side.

Comment (xster) [inline packages/flutter/lib/src/painting/box_painter.dart]: ha, no worries, I kinda held my breath on this one too

**Pull request merged**